File: code/packages/csharp/verilog-lexer/VerilogLexer.cs
using System.Collections.Concurrent;
using System.Text;
using CodingAdventures.GrammarTools;
using CodingAdventures.Lexer;

namespace CodingAdventures.VerilogLexer;

/// <summary>
/// Verilog lexer backed by shared grammar token definitions.
/// </summary>
public static class VerilogLexer
{
    public const string DefaultVersion = "2005";
    public static IReadOnlyList<string> SupportedVersions { get; } = new[] { "1995", "2001", "2005" };

    private static readonly ConcurrentDictionary<string, TokenGrammar> TokenGrammars = new();

    public static GrammarLexer CreateVerilogLexer() => CreateVerilogLexer(DefaultVersion);

    public static GrammarLexer CreateVerilogLexer(string? version) => new(LoadTokenGrammar(version));

    public static IReadOnlyList<Token> TokenizeVerilog(string source) => TokenizeVerilog(source, DefaultVersion);

    public static IReadOnlyList<Token> TokenizeVerilog(string source, string? version)
    {
        try
        {
            return CreateVerilogLexer(version).Tokenize(source);
        }
        catch (LexerError error)
        {
            throw new ArgumentException("Verilog tokenization failed: " + error.Message, nameof(source), error);
        }
    }

    private static TokenGrammar LoadTokenGrammar(string? version)
    {
        var validated = ValidateVersion(version);
        return TokenGrammars.GetOrAdd(validated, ParseTokenGrammarResource);
    }

    private static string ValidateVersion(string? version)
    {
        if (string.IsNullOrWhiteSpace(version))
        {
            return DefaultVersion;
        }

        if (!SupportedVersions.Contains(version))
        {
            throw new ArgumentException("Unknown Verilog version '" + version + "'. Valid values: " + string.Join(", ", SupportedVersions), nameof(version));
        }

        return version;
    }

    private static TokenGrammar ParseTokenGrammarResource(string version)
    {
        try
        {
            return TokenGrammarParser.Parse(ReadResource("verilog" + version + ".tokens"));
        }
        catch (TokenGrammarError error)
        {
            throw new InvalidOperationException("Failed to parse bundled Verilog token grammar for version " + version, error);
        }
    }

    private static string ReadResource(string resourceName)
    {
        var assembly = typeof(VerilogLexer).Assembly;
        using var stream = assembly.GetManifestResourceStream(resourceName)
            ?? throw new InvalidOperationException("Missing bundled resource: " + resourceName);
        using var reader = new StreamReader(stream, Encoding.UTF8);
        return reader.ReadToEnd();
    }
}

File: code/packages/fsharp/verilog-lexer/VerilogLexer.fs
namespace CodingAdventures.VerilogLexer.FSharp

open System
open System.Collections.Concurrent
open System.Collections.Generic
open System.IO
open System.Text
open CodingAdventures.GrammarTools.FSharp
open CodingAdventures.Lexer.FSharp

/// Verilog lexer backed by shared grammar token definitions.
type VerilogLexer private () =
    static let defaultVersion = "2005"
    static let supportedVersions = [| "1995"; "2001"; "2005" |]
    static let tokenGrammars = ConcurrentDictionary<string, TokenGrammar>()

    static member DefaultVersion = defaultVersion
    static member SupportedVersions = supportedVersions :> IReadOnlyList<string>

    static member CreateVerilogLexer() = VerilogLexer.CreateVerilogLexer(defaultVersion)
    static member CreateVerilogLexer(version: string) = GrammarLexer(VerilogLexer.LoadTokenGrammar(version))

    static member TokenizeVerilog(source: string) = VerilogLexer.TokenizeVerilog(source, defaultVersion)
    static member TokenizeVerilog(source: string, version: string) =
        try
            VerilogLexer.CreateVerilogLexer(version).Tokenize(source)
        with
        | :? LexerError as error ->
            raise (ArgumentException(sprintf "Verilog tokenization failed: %s" error.Message, "source", error))

    static member private LoadTokenGrammar(version: string) =
        let validated = VerilogLexer.ValidateVersion(version)
        tokenGrammars.GetOrAdd(validated, Func<string, TokenGrammar>(VerilogLexer.ParseTokenGrammarResource))

    static member private ValidateVersion(version: string) =
        if String.IsNullOrWhiteSpace(version) then
            defaultVersion
        elif Array.exists ((=) version) supportedVersions then
            version
        else
            invalidArg "version" (sprintf "Unknown Verilog version '%s'. Valid values: %s" version (String.Join(", ", supportedVersions)))

    static member private ParseTokenGrammarResource(version: string) =
        try
            TokenGrammarParser.Parse(VerilogLexer.ReadResource("verilog" + version + ".tokens"))
        with
        | :? TokenGrammarError as error ->
            raise (InvalidOperationException(sprintf "Failed to parse bundled Verilog token grammar for version %s" version, error))

    static member private ReadResource(resourceName: string) =
        let assembly = typeof<VerilogLexer>.Assembly
        use stream = assembly.GetManifestResourceStream(resourceName)
        if isNull stream then
            invalidOp (sprintf "Missing bundled resource: %s" resourceName)
        use reader = new StreamReader(stream, Encoding.UTF8)
        reader.ReadToEnd()
